README
Parse-FixedRecord

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Parse::FixedRecord

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-FixedRecord

    Search CPAN
        http://search.cpan.org/dist/Parse-FixedRecord/


COPYRIGHT AND LICENCE

Copyright (C) 2009 osfameron

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Parse-FixedRecord

0.02    2009 Dec 09
        Added missing dep on DateTime::Format::Strptime

0.01    2009 Dec 08
        First version


lib/Parse/FixedRecord/Row.pm
package Parse::FixedRecord::Row;

=head1 NAME

Parse::FixedRecord::Row - Row class for Parse::FixedRecord

=head1 DESCRIPTION

This is the base class for fixed record parsers.
Provides the C<parse> implementation.

=head2 Methods

=head3 C<parse>

See L<Parse::FixedRecord> for usage;

=head3 C<output>

Provides stringified output.  This is overloaded, so you can just:

    print $obj;

to get an output (in the same format as declared/parsed).  This depends
on each individual parser type having well behaved String overloading!

=head2 Types

::Row declares C<Duration> and C<DateTime> types for you to use in your
parsers.

=head1 AUTHOR and LICENSE

    osfameron@cpan.org - see Parse::FixedRecord for full details

=cut

use Parse::FixedRecord;
use Parse::FixedRecord::Column;

use MooseX::ClassAttribute;
use Moose::Util::TypeConstraints;
use DateTime::Format::Strptime;
use DateTime::Format::Duration;
use List::Util qw/max/;

use overload q("") => sub { $_[0]->output };

subtype 'My::MMA' =>
    as class_type('Moose::Meta::Attribute');

class_has fields => (
    metaclass => 'Collection::Array',
    default   => sub { [] },
    is        => 'rw',
    isa       => 'ArrayRef[Str|My::MMA]',
    provides  => {
        push => 'add_field',
        },
    );

subtype 'Date' =>
    as class_type('DateTime');

coerce 'Date'
    => from 'Str'
        => via { 
            eval {
                my $f = DateTime::Format::Strptime->new( pattern => '%F' );
                my $d = $f->parse_datetime( $_ );
                $d->set_formatter($f);
                $d;
            }
            };

subtype 'Duration' =>
    as class_type('DateTime::Duration');

coerce 'Duration'
    => from 'Str'
        => via { 
            my $f = DateTime::Format::Duration->new( pattern => '%R' );
            my $d = $f->parse_duration( $_ );
            $d->{formatter} = $f;                      # direct access! Yuck!
            bless $d, 'DateTime::Duration::Formatted'; # rebless!
            return $d;
            };

sub parse {
    my ($self, $string) = @_;
    my $class = ref $self || $self;

    my @ranges = @{ $class->fields };

    my $pos = 0;
    my %data = map {
        if (ref) {
            # it's an attribute
            my $width = $_->width;
            my $name  = $_->name;
            my $value = substr($string, $pos, $width);
            $pos += $width;
            ($name => $value);
        } else {
            # it's a string
            my $width = length;
            substr($string, $pos, $width) eq $_ or die "Invalid parse on picture '$_' ($pos)";
            $pos += $width;
            ();
        }
        } @ranges;

    return $class->new( %data );
}

sub output {
    my ($self) = @_;

    my @ranges = @{ $self->fields };

    my $string = join '', map {
        if (ref) {
            my $width = $_->width;
            sprintf "\%${width}s", $_->get_value($self);
        } else {
            $_
        }} @ranges;
    return $string;
}

package DateTime::Duration::Formatted;
our @ISA = 'DateTime::Duration';

use overload q("") => sub {
    my ($self) = @_;
    my $f = $self->{formatter};
    return $f->format_duration_from_deltas($f->normalise($self));
    };

1;


lib/Parse/FixedRecord.pm
package Parse::FixedRecord;

=head1 NAME

Parse::FixedRecord - object oriented parser for fixed width records

=head1 SYNOPSIS

Assuming you have data like this:

  Fred Bloggs | 2009-12-08 | 01:05
  Mary Blige  | 2009-12-08 | 00:30

To create a parser:

  package My::Parser;
  use Parse::FixedRecord; # imports strict and warnings
  extends 'Parse::FixedRecord::Row';

  column first_name => width => 4, isa => 'Str';
  pic ' ';
  column last_name  => width => 6, isa => 'Str';
  pic ' | ';
  column date       => width => 10, isa => 'Date';
  pic ' | ';
  column duration   => width => 5, isa => 'Duration';
  1;

In your code:

  use My::Parser;
  while (my $line = <$fh>) {
    eval {
      my $object = My::Parser->parse( $line );
      say $object->first_name;
      do_something() if $ $object->duration->in_units('mins') > 60;
    };
  }

=head1 DESCRIPTION

C<Parse::FixedRecord> is a subclass of L<Moose> with a simple domain specific
language (DSL) to define parsers.

You may use any type constraints you like, as long as they have a coercion
from Str.  If you wish to output row objects in the same format, they should
also have an overload.

C<Parse::FixedRecord> provides C<Duration> and C<DateTime> constraints for
you out of the box.

=head2 Definition

To define the class, simply apply C<column> and C<pic> for each field, in
the order they appear in your input file.  They are defined as follows:

=head3 C<column>

This is a specialisation of C<Moose>'s C<has>, which applies the
L<Parse::FixedRecord::Column> trait.

You must supply a 'width' parameter.
Unless you specify otherwise, the trait will default to C<is =E<gt> 'ro'>,
so will be readonly.

  column foo => width => 10;                     # readonly accessor
  column bar => width => 5, is => 'rw';          # read/write
  column baz => width => 5, isa => 'Some::Type';

=head3 C<pic>

You may also supply delimiters.  As this is a fixed record parser, allowing
delimiters may seem odd.  But it may be convenient for some (odd) datasets,
and in any case, there is no requirement to use it.

  column foo => width => 5;
  pic ' | ';
  column bar => width => 5;

i.e. the record consists of two 5-char wide fields, split by the literal 
C<' | '>.

=head2 Parsing

=head3 C<$parser-E<gt>parse>

  my $obj = My::Parser->parse( $line );

If the C<column> and C<pic> definitions can be matched, including any
type constraints and object inflations, then a Moose object is returned.

Otherwise, an error is thrown, usually by the Moose type constraint failure.

=cut

use Moose ();
use Parse::FixedRecord::Column;
use Moose::Exporter;
use Moose::Util::TypeConstraints;

our $VERSION = 0.02;

Moose::Exporter->setup_import_methods(
   with_caller => ['column', 'pic'],
   also        => ['Moose' ],
);

sub pic {
    my $caller = shift;
    my $pic = shift;

    $caller->add_field($pic);
}

sub column {
    my $caller = shift;
    my ($name, %pars) = @_;
    $pars{isa} ||= 'Str';
    $pars{coerce}++ if do {
        my $t = find_type_constraint($pars{isa});
        $t && $t->has_coercion;
        };
    my $attr = $caller->meta->add_attribute(
        $name => (
            traits => ['Column'],
            is     => 'ro',
            %pars,
            ));
    $caller->add_field($attr);
}

=head1 AUTHOR and LICENSE

   (C)  osfameron 2009, <osfameron@cpan.org>

For support, try emailing me, or grabbing me on irc #london.pm or #moose
on irc.perl.org

This module is released under the same terms as Perl itself.

=cut

1;
